### ISSUE TITLE
Big performance drop at runtime from 1.29 to 1.34

### ISSUE THREAD

**Issue by warrenseine** [kripken/emscripten#3684]
We're seeing a massive performance drop (from 60 FPS to 15 FPS) when upgrading to 1.34 (`sdk-master-64bit`) in an OpenGL application.

I've made no code change and there's no visible error in the console. It's a private application so I can't share it here unfortunately.

Before I dig into details, do you have any idea what might have changed to impact performances since 1.29? A change in the main loop? An SDL update?

**Comment by kripken** [comment]
That's very surprising. Nothing I can think of could explain that.

The simplest approach here might be to bisect over versions of emscripten, if you find that `1.x.y` is ok but `1.x.y+1` is  bad, that should give us enough information.

**Comment by kvantor** [comment]
warrenseine, why are you can't compare changelog with your old version and give diff results here?

**Comment by juj** [comment]
Such a large performance drop should likely show up in the browser profilers as well? If bisecting doesn't find anything, try taking performance profiles in FF Dev Edition of both versions and cross-reference those?

**Comment by warrenseine** [comment]
I went back to 1.29 for some time. Now, back to understanding the issue.

**Comment by juj** [comment]
I think this issue is an important one we should keep a track of before next release. I am not sure how Unity3D or Unreal Engine 4 are affected, if at all, but dahlia on IRC pointed out that Bullet physics in the open source Urho3D is affected by a performance bug like this (Firefox retains 60fps, but in Chrome performance regressed from 60fps to 5fps). Profiling showed a standard 3D dot product as the hottest function in the profile. This was a standard `-s PRECISE_F32=0` build. I tried to do a custom build of Urho3D physics samples to bisect a while back, but ran into some build system issues there and have not yet had time yet to find the offending commit. Enabling memory growth did cause Chrome perf issues I recall, but iirc the Urho3D build did not have that enabled.

**Comment by kripken** [comment]
Does bullet in the benchmark suite show this regression?

**Comment by kripken** [comment]
In the benchmark suite, I see Chrome 1.69x slower than Firefox, which is about where it's been for a long time, so it doesn't seem like any big regression is noticeable there.

`PRECISE_F32` and memory growth are generally slow on Chrome, so either of those could have explained it, but you ruled that out; also they would not have changed by Emscripten version.

If you are busy, let me know how to build Urho and reproduce this issue, I might have time to investigate.

**Comment by juj** [comment]
After the fix https://github.com/kripken/emscripten/pull/3819 , one can do a build of Urho3D with

	git clone https://github.com/urho3d/Urho3D.git
	cd Urho3D
	mkdir build_incoming
	./cmake_emscripten.sh build_incoming -DURHO3D_SAMPLES=1 -DCMAKE_BUILD_TYPE=RelWithDebInfo
	cd build_incoming
	make -j8

Then the samples are outputted to the directory `bin/`. However building with PR #3819 applied on top of latest incoming, I do not see a performance impact on the Urho3D test case 12_PhysicsStressTest. When I ran versions of that sample that dahlia had built with new and old Emscripten versions, I was able to see the performance regression there. This means that either the issue has been resolved somehow, or that my builds have some difference that did not reproduce the issue.

I won't close this bug though, since I jumped in on this with a guess that this sample would have provided a repro case to @warrenseine 's issue, but unfortunately it did not, and his issue might be different.
